numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -816,66 +816,54 @@ halftype_@kind@(PyObject *self)
 
 /**end repeat**/
 
-
-/*
- * Could improve this with a PyLong_FromLongDouble(longdouble ldval)
- * but this would need some more work...
- */
-
 /**begin repeat
- *
- * #name = (int, float)*2#
- * #KIND = (Long, Float)*2#
- * #char = ,,c*2#
- * #CHAR = ,,C*2#
- * #POST = ,,.real*2#
+ * #char = ,c#
+ * #CHAR = ,C#
+ * #POST = ,.real#
  */
 static PyObject *
-@char@longdoubletype_@name@(PyObject *self)
+@char@longdoubletype_float(PyObject *self)
 {
-    double dval;
-    PyObject *obj, *ret;
+    npy_longdouble val = PyArrayScalar_VAL(self, @CHAR@LongDouble)@POST@;
+    return PyFloat_FromDouble((double) val);
+}
 
-    dval = (double)(((Py@CHAR@LongDoubleScalarObject *)self)->obval)@POST@;
-    obj = Py@KIND@_FromDouble(dval);
-    if (obj == NULL) {
-        return NULL;
-    }
-    ret = Py_TYPE(obj)->tp_as_number->nb_@name@(obj);
-    Py_DECREF(obj);
-    return ret;
+static PyObject *
+@char@longdoubletype_long(PyObject *self)
+{
+    npy_longdouble val = PyArrayScalar_VAL(self, @CHAR@LongDouble)@POST@;
+
+    /*
+     * This raises OverflowError when the cast overflows to infinity (gh-9964)
+     *
+     * Could fix this with a PyLong_FromLongDouble(longdouble ldval)
+     * but this would need some more work...
+     */
+    return PyLong_FromDouble((double) val);
```

Comment:
This is where things need to change to fix #9964, but lets leave that to a future PR.